crates/bevy_render/src/renderer/render_device.rs

```diff
@@ -45,7 +45,11 @@ impl RenderDevice {
     /// Creates a [`ShaderModule`](wgpu::ShaderModule) from either SPIR-V or WGSL source code.
     #[inline]
     pub fn create_shader_module(&self, desc: wgpu::ShaderModuleDescriptor) -> wgpu::ShaderModule {
-        self.device.create_shader_module(desc)
+        if cfg!(target_arch = "wasm32") {
+            self.device.create_shader_module(desc)
+        } else {
+            unsafe { self.device.create_shader_module_unchecked(desc) }
```

Comment:
The [safety invariant](https://docs.rs/wgpu/latest/wgpu/struct.Device.html#safety) is not validated by the code here, or is it exposed as an unsafe invariant to the caller. This would be considered unsound.

Reply:
On the one hand, this would disable soundness checks like out of bounds reads. On the other, it is still very easy to pass garbage data in bindings that would cause practically out of bounds reads because there is garbage data within the bounds of the bindings, but outside the bounds of what is meant to be used. We can fix that too with runtime checks in shaders, but it would be by convention for code that we write in core and it would have a performance impact.

Still, I feel like there could be a desire for an unsafe feature for rendering that disables runtime checks for more performance. wgpu does a lot of runtime checks, but for example, Vulkan manages the checks through 'validation layers' which are intended to be enabled during development to identify invalid API usage, and then disabled at production runtime for performance. I could see this tradeoff being a desirable choice for some. I don't know if wgpu supports disabling runtime checks though.

Also, the documentation says that it has no impact on web.

Reply:
> Still, I feel like there could be a desire for an unsafe feature for rendering that disables runtime checks for more performance. wgpu does a lot of runtime checks, but for example, Vulkan manages the checks through 'validation layers' which are intended to be enabled during development to identify invalid API usage, and then disabled at production runtime for performance. I could see this tradeoff being a desirable choice for some. I don't know if wgpu supports disabling runtime checks though.

If this is the case, we should expose a wrapper around `create_shader_module_unchecked` as an unsafe API, and expose the same invariants to the caller, not wrap it in a safe function and assume that users will not call it with garbage inputs.

I'm also not 100% in agreement that we should disable it in release mode as a feature. Shaders and materials are assets in Bevy, stored and loaded from disk or the network. Inherently trusting them and yeeting them into the GPU without safety checks seems like a great way to cause driver crashes, or, in the worst case, potential security issues.